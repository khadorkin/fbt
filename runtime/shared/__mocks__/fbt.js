/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * @format
 * @emails oncall+internationalization
 */

jest
  .unmock('FbtEnv')
  .unmock('FbtTable')
  .unmock('FbtTableAccessor')
  .unmock('intlNumUtils')
  .unmock('IntlPhonologicalRewrites')
  .unmock('IntlPunctuation')
  .unmock('IntlVariationResolver')
  .unmock('IntlVariationResolverImpl')
  .unmock('NumberFormatConsts')
  .unmock('substituteTokens')
  .mock('FbtNumberType');

const fbt = jest.fn();
const fbtRuntime = jest.requireActual('fbt');

const WRAPPER = '__FBT__';

fbt.mockImplementation(function () {
  throw new Error('should never be called');
});

function unwrap(json) {
  try {
    return JSON.parse(json.slice(WRAPPER.length, -WRAPPER.length));
  } catch (ex) {
    throw new Error('Invalid FBT JSON: ' + json + ' (' + ex.message + ')');
  }
}

fbt._ = jest.fn();
fbt._.mockImplementation(function (wrappedJSON, args) {
  const unwrappedJson = unwrap(wrappedJSON);
  const jsfbt = unwrappedJson.jsfbt;
  return fbtRuntime._(unwrappedJson.type === 'text' ? jsfbt : jsfbt.t, args);
});

fbt._.getCallString = function (index) {
  return unwrap(fbt._.mock.calls[index][0]);
};

fbt._param = function (name, value, variations) {
  return fbtRuntime._param(name, value, variations);
};

fbt._plural = function (count, label, value) {
  return fbtRuntime._plural(count, label, value);
};

fbt._pronoun = function (usage, gender, options) {
  return fbtRuntime._pronoun(usage, gender, options);
};

fbt._enum = function (value, range) {
  return fbtRuntime._enum(value, range);
};

fbt._subject = function (value) {
  return fbtRuntime._subject(value);
};

fbt._name = function (...whateveryo) {
  return fbtRuntime._name(...whateveryo);
};

fbt.isFbtInstance = fbtRuntime.isFbtInstance;

module.exports = fbt;
