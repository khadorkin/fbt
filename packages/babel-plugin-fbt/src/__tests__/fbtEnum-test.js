/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * @emails oncall+i18n_fbt_js
 * @format
 */

jest.autoMockOff();

const TestFbtEnumManifest = require('TestFbtEnumManifest');

const {
  jsCodeFbtCallSerializer,
  snapshotTransform,
  withFbtRequireStatement,
} = require('../FbtTestUtil');

expect.addSnapshotSerializer(jsCodeFbtCallSerializer);

function runTest(data) {
  expect(
    snapshotTransform(data.input, {fbtEnumManifest: TestFbtEnumManifest}),
  ).toMatchSnapshot();
}

// TODO(T40113359) Re-enable once this test scenario is ready to be tested
xdescribe('Test Fbt Enum', () => {
  beforeEach(() => {
    // Ensure the Enum registrar config is reset.
    jest.resetModules();
  });

  it('should handle jsx enums (with references)', () => {
    runTest({
      input: withFbtRequireStatement(
        `let aEnum = require('Test$FbtEnum');
          var x = (
            <fbt desc="enums!">
              Click to see
              <fbt:enum enum-range={aEnum} value={id} />
            </fbt>
          );`,
      ),
    });
  });

  it('should handle jsx string literals', () => {
    runTest({
      input: withFbtRequireStatement(
        `let aEnum = require('Test$FbtEnum');
          var x = (
            <fbt desc="enums!">
              Click to see
              <fbt:enum enum-range={aEnum} value="id1" />
            </fbt>
          );`,
      ),
    });
  });

  it('should handle functional enums (with references) (require)', () => {
    runTest({
      input: withFbtRequireStatement(
        `let aEnum = require('Test$FbtEnum');
          var x = fbt('Click to see ' + fbt.enum(id, aEnum), 'enums!');`,
      ),
    });
  });

  it('should handle functional enums (with references) (import default)', () => {
    runTest({
      input: `
        import fbt from 'fbt';
        import aEnum from 'Test$FbtEnum';
        var x = fbt('Click to see ' + fbt.enum(id, aEnum), 'enums!');
      `,
    });
  });

  it('should handle functional enums (with references) (import star)', () => {
    runTest({
      input: `
        import fbt from 'fbt';
        import * as aEnum from 'Test$FbtEnum';
        var x = fbt('Click to see ' + fbt.enum(id, aEnum), 'enums!');
      `,
    });
  });

  it('should handle functional enums (with references) in templates', () => {
    runTest({
      input: withFbtRequireStatement(
        `let aEnum = require('Test$FbtEnum');
          var x = fbt(\`Click to see \${fbt.enum(id, aEnum)}\`, 'enums!');`,
      ),
    });
  });

  it('should throw when enum values are not strings', () => {
    expect(() =>
      snapshotTransform(
        withFbtRequireStatement(
          `let aEnum = require('Test$FbtEnum');
          var x = fbt('This is ' + fbt.enum(id, {bad: \`egg\`}), 'enums!');`,
        ),
        {fbtEnumManifest: TestFbtEnumManifest},
      ),
    ).toThrowError(
      'fbt enum range values must be StringLiteral, got TemplateLiteral',
    );
  });

  it('should throw on multiple import types', () => {
    expect(() =>
      snapshotTransform(
        withFbtRequireStatement(
          `import aEnum, * as bEnum from 'Test$FbtEnum';
          var x = fbt('Click to see ' + fbt.enum(id, aEnum), 'enums!');`,
        ),
      ),
    ).toThrowError('Fbt Enum `aEnum` not registered');
  });

  it('should throw on destructured imports', () => {
    expect(() =>
      snapshotTransform(
        withFbtRequireStatement(
          `import {aEnum} from 'Test$FbtEnum';
          var x = fbt('Click to see ' + fbt.enum(id, aEnum), 'enums!');`,
        ),
      ),
    ).toThrowError('Fbt Enum `aEnum` not registered');
  });
});
