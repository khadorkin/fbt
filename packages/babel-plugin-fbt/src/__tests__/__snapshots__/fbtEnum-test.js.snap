// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Fbt Enum should handle functional enums (with references) (import default) 1`] = `
import fbt from "fbt";
import aEnum from "Test$FbtEnum";

var x = fbt._(
  /* __FBT__ start */ {
    type: "table",
    jsfbt: {
      t: {
        id1: "Click to see groups",
        id2: "Click to see photos",
        id3: "Click to see videos",
      },
      m: [null],
    },
    desc: "enums!",
    project: "",
  } /* __FBT__ end */,
  [fbt._enum(id, aEnum)]
);

`;

exports[`Test Fbt Enum should handle functional enums (with references) (import star) 1`] = `
import fbt from "fbt";
import * as aEnum from "Test$FbtEnum";

var x = fbt._(
  /* __FBT__ start */ {
    type: "table",
    jsfbt: {
      t: {
        id1: "Click to see groups",
        id2: "Click to see photos",
        id3: "Click to see videos",
      },
      m: [null],
    },
    desc: "enums!",
    project: "",
  } /* __FBT__ end */,
  [fbt._enum(id, aEnum)]
);

`;

exports[`Test Fbt Enum should handle functional enums (with references) (require) 1`] = `
const fbt = require("fbt");

let aEnum = require("Test$FbtEnum");

var x = fbt._(
  /* __FBT__ start */ {
    type: "table",
    jsfbt: {
      t: {
        id1: "Click to see groups",
        id2: "Click to see photos",
        id3: "Click to see videos",
      },
      m: [null],
    },
    desc: "enums!",
    project: "",
  } /* __FBT__ end */,
  [fbt._enum(id, aEnum)]
);

`;

exports[`Test Fbt Enum should handle functional enums (with references) in templates 1`] = `
const fbt = require("fbt");

let aEnum = require("Test$FbtEnum");

var x = fbt._(
  /* __FBT__ start */ {
    type: "table",
    jsfbt: {
      t: {
        id1: "Click to see groups",
        id2: "Click to see photos",
        id3: "Click to see videos",
      },
      m: [null],
    },
    desc: "enums!",
    project: "",
  } /* __FBT__ end */,
  [fbt._enum(id, aEnum)]
);

`;

exports[`Test Fbt Enum should handle jsx enums (with references) 1`] = `
const fbt = require("fbt");

let aEnum = require("Test$FbtEnum");

var x = fbt._(
  /* __FBT__ start */ {
    type: "table",
    jsfbt: {
      t: {
        id1: "Click to see groups",
        id2: "Click to see photos",
        id3: "Click to see videos",
      },
      m: [null],
    },
    desc: "enums!",
    project: "",
  } /* __FBT__ end */,
  [fbt._enum(id, aEnum)]
);

`;

exports[`Test Fbt Enum should handle jsx string literals 1`] = `
const fbt = require("fbt");

let aEnum = require("Test$FbtEnum");

var x = fbt._(
  /* __FBT__ start */ {
    type: "table",
    jsfbt: {
      t: {
        id1: "Click to see groups",
        id2: "Click to see photos",
        id3: "Click to see videos",
      },
      m: [null],
    },
    desc: "enums!",
    project: "",
  } /* __FBT__ end */,
  [fbt._enum("id1", aEnum)]
);

`;
